crates/miden-lib/src/utils/script_builder.rs

```diff
@@ -0,0 +1,458 @@
+use alloc::{string::String, sync::Arc};
+
+use miden_objects::{
+    assembly::{
+        Assembler, Library, LibraryPath,
+        diagnostics::{NamedSource, SourceManager},
+    },
+    note::NoteScript,
+    transaction::TransactionScript,
+};
+
+use crate::{errors::ScriptBuilderError, transaction::TransactionKernel};
+
+// SCRIPT BUILDER
+// ================================================================================================
+
+/// A builder for compiling note scripts and transaction scripts with optional library dependencies.
+///
+/// The ScriptBuilder simplifies the process of creating transaction scripts by providing:
+/// - A clean API for adding multiple libraries with static or dynamic linking
+/// - Automatic assembler configuration with all added libraries
+/// - Debug mode support
+/// - Builder pattern support for method chaining
+///
+/// ## Static vs Dynamic Linking
+///
+/// **Static Linking** (`link_static_library()` / `with_statically_linked_library()`):
+/// - Use when you control and know the library code
+/// - The library code is copied into the script code
+/// - Best for most user-written libraries and dependencies
+/// - Results in larger script size but ensures the code is always available
+///
+/// **Dynamic Linking** (`link_dynamic_library()` / `with_dynamically_linked_library()`):
+/// - Use when making Foreign Procedure Invocation (FPI) calls
+/// - The library code is available on-chain and referenced, not copied
+/// - Best for calling procedures in foreign accounts via FPI
+/// - Results in smaller script size but requires the code to be available on-chain
+///
+/// ## Typical Workflow
+///
+/// 1. Create a new ScriptBuilder with debug mode preference
+/// 2. Add any required modules using `link_module()` or `with_linked_module()`
+/// 3. Add libraries using `link_static_library()` / `link_dynamic_library()` as appropriate
+/// 4. Compile your script with `compile_note_script()` or `compile_tx_script()`
+///
+/// Note that the compilation methods consume the ScriptBuilder, so if you need to compile
+/// multiple scripts with the same configuration, you should clone the builder first.
+///
+/// ## Builder Pattern Example
+///
+/// ```ignore
```

Comment:
Here, it would be nice to use `no_run` instead of `ignore` so that the example is built (but not executed) if the API somehow changes, we can make sure the example is still compatible. You can hide some setup lines with `#` so that they get built but not rendered in the docs.